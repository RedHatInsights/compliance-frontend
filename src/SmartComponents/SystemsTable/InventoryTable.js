import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { useStore } from 'react-redux';
import { Alert, Spinner } from '@patternfly/react-core';
import { TableVariant } from '@patternfly/react-table';
// eslint-disable-next-line max-len
import ComplianceRemediationButton from '@redhat-cloud-services/frontend-components-inventory-compliance/ComplianceRemediationButton';
import { exportFromState } from 'Utilities/Export';
import { DEFAULT_SYSTEMS_FILTER_CONFIGURATION, COMPLIANT_SYSTEMS_FILTER_CONFIGURATION } from '@/constants';
import { ErrorPage, StateView, StateViewPart } from 'PresentationalComponents';
import { systemsWithRuleObjectsFailed } from 'Utilities/ruleHelpers';
import useFilterConfig from 'Utilities/hooks/useFilterConfig';
import { InventoryTable as FECInventoryTable } from '@redhat-cloud-services/frontend-components/Inventory';
import { policyFilter, defaultOnLoad } from './constants';
import {
    useFetchSystems, useGetEntities, useOsMinorVersionFilter, useInventoryUtilities, useOnSelect
} from './hooks';

export const InventoryTable = ({
    columns,
    showAllSystems,
    policyId,
    query,
    showActions,
    enableExport,
    compliantFilter,
    policies,
    showOnlySystemsWithTestResults,
    showOsFilter,
    error,
    showComplianceSystemsInfo,
    compact,
    remediationsEnabled,
    systemProps,
    defaultFilter,
    emptyStateComponent,
    prependComponent,
    showOsMinorVersionFilter,
    preselectedSystems,
    onSelect: onSelectProp
}) => {
    const store = useStore();
    const inventory = useRef(null);
    const [isEmpty, setIsEmpty] = useState(false);
    const [isLoaded, setIsLoaded] = useState(false);
    const [items, setItems] = useState([]);
    const [total, setTotal] = useState(0);

    const {
        onSelect, onBulkSelect, selectedSystems, isPageSelected
    } = useOnSelect(onSelectProp, items, preselectedSystems, total);
    const selectedIds = selectedSystems?.map((system) => (system.id));
    const selectedItems = items.filter((item) => (selectedIds.includes(item.id)));
    const selectedCount = selectedSystems.length;

    const osMinorVersionFilter = useOsMinorVersionFilter(showOsMinorVersionFilter);
    const { conditionalFilter, activeFilters, buildFilterString, activeFilterValues } = useFilterConfig([
        ...DEFAULT_SYSTEMS_FILTER_CONFIGURATION,
        ...(compliantFilter ? COMPLIANT_SYSTEMS_FILTER_CONFIGURATION : []),
        ...(policies?.length > 0 ? policyFilter(policies, showOsFilter) : []),
        ...osMinorVersionFilter
    ]);

    useInventoryUtilities(inventory, selectedSystems, activeFilterValues);

    const onComplete = (result) => {
        setTotal(result.meta.totalCount);
        setItems(result.entities);
        setIsLoaded(true);

        if (emptyStateComponent &&
            result.meta.totalCount === 0 &&
            activeFilterValues.length === 0) {
            setIsEmpty(true);
        }
    };

    const fetchSystems = useFetchSystems(
        query, policyId, buildFilterString, showOnlySystemsWithTestResults, defaultFilter, onComplete
    );
    const getEntities = useGetEntities(fetchSystems, { selected: selectedSystems });

    return <StateView stateValues={{ error, noError: error === undefined && !isEmpty, empty: isEmpty }}>
        <StateViewPart stateKey='error'>
            { !!prependComponent && prependComponent }
            <ErrorPage error={error}/>
        </StateViewPart>
        <StateViewPart stateKey='empty'>
            { emptyStateComponent }
        </StateViewPart>
        <StateViewPart stateKey='noError'>
            { !!prependComponent && isLoaded && prependComponent }
            { showComplianceSystemsInfo && <Alert
                isInline
                variant="info"
                title={ 'The list of systems in this view is different than those that appear in the Inventory. ' +
                    'Only systems currently associated with or reporting against compliance policies are displayed.' } /> }
            <FECInventoryTable
                { ...systemProps }
                ref={ inventory }
                activeFilters={ activeFilters }
                getEntities={ getEntities }
                onLoad={ defaultOnLoad(columns) }
                hideFilters={{
                    name: true,
                    tags: true,
                    registeredWith: true,
                    stale: true
                }}
                tableProps={{
                    canSelectAll: false,
                    ...items.length > 0 && { onSelect }
                }}
                fallback={ <Spinner /> }
                variant={ compact ? TableVariant.compact : '' }
                bulkSelect={{
                    checked: selectedCount > 0 ? (isPageSelected ? true : null) : false,
                    onSelect: items.length > 0 && onBulkSelect,
                    count: selectedCount,
                    label: selectedCount > 0 ? `${ selectedCount } Selected` : undefined
                }}
                {...!showAllSystems && {
                    ...conditionalFilter,
                    ...remediationsEnabled && {
                        dedicatedAction: <ComplianceRemediationButton
                            allSystems={ systemsWithRuleObjectsFailed(selectedItems) }
                            selectedRules={ [] } />
                    }
                }}
                {...enableExport && {
                    exportConfig: {
                        isDisabled: total === 0 && selectedCount === 0,
                        onSelect: (_, format) => exportFromState(store.getState()?.entities, format)
                    }
                }}
                {...showActions && {
                    actions: [{
                        title: 'View in inventory',
                        onClick: (_event, _index, { id }) => {
                            const beta = window.location.pathname.split('/')[1] === 'beta';
                            window.location.href = `${window.location.origin}${beta ? '/beta' : ''}/insights/inventory/${id}`;
                        }
                    }]
                }}
            />
        </StateViewPart>
    </StateView>;
};

InventoryTable.propTypes = {
    columns: PropTypes.arrayOf(PropTypes.shape({})),
    policies: PropTypes.arrayOf(PropTypes.shape({})),
    showAllSystems: PropTypes.bool,
    policyId: PropTypes.string,
    query: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.object
    ]),
    showActions: PropTypes.bool,
    enableExport: PropTypes.bool,
    compliantFilter: PropTypes.bool,
    showOnlySystemsWithTestResults: PropTypes.bool,
    showOsFilter: PropTypes.bool,
    showComplianceSystemsInfo: PropTypes.bool,
    error: PropTypes.object,
    compact: PropTypes.bool,
    remediationsEnabled: PropTypes.bool,
    defaultFilter: PropTypes.string,
    systemProps: PropTypes.shape({
        isFullView: PropTypes.bool
    }),
    emptyStateComponent: PropTypes.node,
    prependComponent: PropTypes.node,
    showOsMinorVersionFilter: PropTypes.oneOfType([
        PropTypes.bool,
        PropTypes.arrayOf(PropTypes.string)
    ]),
    preselectedSystems: PropTypes.array,
    onSelect: PropTypes.func
};

InventoryTable.defaultProps = {
    policyId: '',
    showActions: true,
    enableExport: true,
    compliantFilter: false,
    showOnlySystemsWithTestResults: false,
    showComplianceSystemsInfo: false,
    compact: false,
    remediationsEnabled: true,
    preselectedSystems: []
};

export default InventoryTable;
