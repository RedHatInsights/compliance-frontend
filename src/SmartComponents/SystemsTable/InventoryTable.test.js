import { init } from 'Store';
import logger from 'redux-logger';
import { useSelector } from 'react-redux';

import { InventoryTable } from './InventoryTable';

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn(),
    useStore: jest.fn(),
    useDispatch: jest.fn()
}));

const items = {
    data: {
        systems: {
            totalCount: 1,
            edges: [
                {
                    node: {
                        id: 1
                    }
                }
            ]
        }
    }
};

describe('InventoryTable', () => {
    const store = init(logger).getStore();
    const client = { query: jest.fn(() => Promise.resolve(items)) };
    const updateSystemsFunction = jest.fn(jest.fn(() => Promise.resolve(items)));
    const updateRowsFunction = jest.fn(jest.fn(() => Promise.resolve(items)));
    const defaultProps = {
        store,
        client,
        updateSystems: updateSystemsFunction,
        updateRows: updateRowsFunction,
        clearInventoryFilter: jest.fn(),
        titleComponent: 'Title Component',
        emptyStateComponent: 'Empty State Component'
    };
    const MockComponent = jest.fn(({ children, loaded }) => {
        return children && loaded ? children : 'Loading...';
    });

    beforeEach(() => {
        global.insights = {
            loadInventory: jest.fn(() => {
                return Promise.resolve({
                    inventoryConnector: () => ({
                        InventoryTable: MockComponent
                    }),
                    INVENTORY_ACTION_TYPES: {},
                    mergeWithEntities: () => ({})
                });
            })
        };

    });

    it('expect to not render a loading state', () => {
        const state = {
            entities: {
                selectedEntities: []
            }
        };
        useSelector.mockImplementation((callback) => (callback(state)));
        const wrapper = shallow(
            <InventoryTable { ...defaultProps } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
