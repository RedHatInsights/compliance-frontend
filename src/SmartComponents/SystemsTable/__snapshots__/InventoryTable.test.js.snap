// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryTable expect to not render a loading state 1`] = `
<StateView
  stateValues={
    Object {
      "empty": false,
      "error": undefined,
      "noError": true,
    }
  }
>
  <StateViewPart
    stateKey="error"
  >
    <ErrorPage />
  </StateViewPart>
  <StateViewPart
    stateKey="empty"
  >
    Empty State Component
  </StateViewPart>
  <StateViewPart
    stateKey="noError"
  >
    <ForwardRef
      actions={
        Array [
          Object {
            "onClick": [Function],
            "title": "View in inventory",
          },
        ]
      }
      activeFiltersConfig={
        Object {
          "filters": Array [],
          "onDelete": [Function],
        }
      }
      bulkSelect={
        Object {
          "checked": false,
          "count": 0,
          "label": undefined,
          "onSelect": [Function],
        }
      }
      dedicatedAction={
        <Memo(Connect(ComplianceRemediationButton))
          allSystems={Array []}
          selectedRules={Array []}
        />
      }
      exportConfig={
        Object {
          "isDisabled": true,
          "onSelect": [Function],
        }
      }
      fallback={
        <SkeletonTable
          canSelectAll={false}
          colSize={2}
          hasRadio={false}
          isSelectable={false}
          paddingColumnSize={0}
          rowSize={15}
          variant={null}
        />
      }
      filterConfig={
        Object {
          "items": Array [
            Object {
              "filterValues": Object {
                "onChange": [Function],
                "value": "",
              },
              "id": "name",
              "label": "Name",
              "placeholder": "Filter by name",
              "type": "text",
            },
          ],
        }
      }
      isLoaded={false}
      items={Array []}
      onLoad={[Function]}
      onRefresh={[Function]}
      page={1}
      perPage={50}
      tableProps={
        Object {
          "canSelectAll": false,
        }
      }
      total={0}
      variant=""
    />
  </StateViewPart>
</StateView>
`;
